vivisect/__init__.py

```diff
@@ -2789,6 +2802,73 @@ def loadFromMemory(self, memobj, baseaddr, fmtname=None):
         # Snapin our analysis modules
         self._snapInAnalysisModules()
 
+    def writeMemory(self, va, bytez):
+        '''
+        Override writeMemory to hook into the Event subsystem.
+        '''
+        fname = self.getFileByVa(va)
+        fileva = self.getFileMeta(fname, 'imagebase')
+        off = va - fileva
+        self._fireEvent(VWE_WRITEMEM, (fname, off, bytez, self._supervisor))
+
+    def connectImportsWithExports(self):
+        """
+        Look for any "imported" symbols that are satisfied by current exports.
+        Wire up the connection.
+
+        Currently this is simply a pointer write at the location of the Import.
+        If this behavior is ever insufficient, we'll want to track the special
+        nature through the Export/Import events.
+        """
+        logger.info('linking Imports with Exports')
+        # store old setting and set _supervisor mode (so we can write wherever
+        # we want, regardless of permissions)
+        oldsup = self._supervisor
+        self._supervisor = True
+
+        for iva, isz, itype, isym in self.getImports():
+            impfname, impsym = isym.split('.', 1)
+            for eva, num, esym, efname in self.getExports():
+                if impsym != esym:
+                    continue
+
+                if impfname not in (efname, '*'):
+                    continue
+
+                if self.isFunctionThunk(eva):
+                    logger.info("Skipping Exported Thunk")
+                    continue
+
+                # file and symbol name match.  apply the magic.
+                # do we ever *not* write in the full address at the import site?
+                logger.debug("connecting Import 0x%x -> Export 0x%x (%r)", iva, eva, isym)
+                self.writeMemoryPtr(iva, eva)
+
+                # remove the LOC_IMPORT and make it a Pointer instead
+                self.delLocation(iva)
```

Comment:
fair.  i've been having the same thoughts as i have been working with it.  
the emutils usage of this was benefited by deleting it, but i'll rework that.  i think not deleting it is probably valuable.

Reply:
this does have the unfortunate effect of making two locations at the same VA, which still has the same undesired effect.  i think it may be time to track IMPORTS as more than just a location, like we do with EXPORTS.

there are several reasons why this has to become a pointer.
one of the problems comes while attempting to traverse between the linked maps.  others occur during non-funconly emulation.

i thought i could do this without deleting the IMPORT location, but i'm now of the opinion that we must.

Reply:
@rakuy0  more thoughts?